File: src/main/java/com/chattriggers/ctjs/triggers/OnRenderTrigger.java
package com.chattriggers.ctjs.triggers;

import net.minecraft.client.renderer.GlStateManager;
import net.minecraftforge.client.event.RenderGameOverlayEvent;

public class OnRenderTrigger extends OnTrigger {
    private boolean triggerIfCanceled;

    public OnRenderTrigger(Object method, TriggerType triggerType) {
        super(method, triggerType);
        triggerIfCanceled = true;
    }

    /**
     * Sets if the render trigger should run if the event has already been canceled.
     * True by default.
     * @param bool Boolean to set
     * @return the trigger object for method chaining
     */
    public OnRenderTrigger triggerIfCanceled(boolean bool) {
        this.triggerIfCanceled = bool;
        return this;
    }

    @Override
    public void trigger(Object... args) {
        if (!(args[0] instanceof RenderGameOverlayEvent))
            throw new IllegalArgumentException("Argument 0 must be a RenderGameOverlayEvent");
        RenderGameOverlayEvent event = (RenderGameOverlayEvent) args[0];

        if (!triggerIfCanceled && event.isCanceled()) return;

        GlStateManager.pushMatrix();
        callMethod(args);
        GlStateManager.popMatrix();
    }
}

File: src/main/kotlin/com/chattriggers/ctjs/triggers/OnRenderTrigger.kt
package com.chattriggers.ctjs.triggers

import com.chattriggers.ctjs.engine.ILoader
import com.chattriggers.ctjs.minecraft.listeners.CancellableEvent
import com.chattriggers.ctjs.utils.kotlin.External

@External
class OnRenderTrigger(method: Any, triggerType: TriggerType, loader: ILoader) : OnTrigger(method, triggerType, loader) {
    private var triggerIfCanceled: Boolean = true

    /**
     * Sets if the render trigger should run if the event has already been canceled.
     * True by default.
     * @param bool Boolean to set
     * @return the trigger object for method chaining
     */
    fun triggerIfCanceled(bool: Boolean) = apply { triggerIfCanceled = bool }

    override fun trigger(args: Array<out Any?>) {
        if (args[0] !is CancellableEvent)
            throw IllegalArgumentException("Argument 0 must be a RenderGameOverlayEvent")

        val event = args[0] as CancellableEvent
        if (!triggerIfCanceled && event.isCanceled())
            return

        callMethod(args)
    }
}
